server/routes/users.js

```diff
@@ -1,11 +1,88 @@
 const express = require('express');
 const router = express.Router();
+const bcrypt = require('bcryptjs');
+const jwt = require('jsonwebtoken');
+const config = require('config');
+const {
+	body,
+	validationResult,
+} = require('express-validator');
+
+const User = require('../models/User');
 
 // @route   POST api/users
 // @desc    Register a user
 // @access  Public
-router.post('/', (req, res) => {
-	res.send('Register a user');
-});
+router.post(
+	'/',
+	[
+		body('firstname', 'Please include your first name')
+			.not()
+			.isEmpty(),
+		body('lastname', 'Please include your last name')
+			.not()
+			.isEmpty(),
+		body('email', 'Please include a valid email'),
+		body(
+			'password',
+			'Please include a password with 6 or more characters'
+		).isLength({ min: 6 }),
+	],
+	async (req, res) => {
+		const errors = validationResult(req);
+		if (!errors.isEmpty()) {
+			return res
+				.status(400)
+				.json({ errors: errors.array() });
+		}
+
+		const { firstname, lastname, email, password } =
+			req.body;
+
+		try {
+			let user = await User.findOne({ email: email });
+
+			if (user) {
+				return res
+					.status(400)
+					.json({ msg: 'Email already in use' });
+			}
+
+			user = new User({
+				firstname,
+				lastname,
+				email,
+				password,
+			});
+
+			const salt = await bcrypt.genSalt(10);
+
+			user.password = await bcrypt.hash(password, salt);
+
+			await user.save();
+
+			const payload = {
+				user: {
+					id: user.id,
+				},
+			};
+
+			jwt.sign(
+				payload,
+				config.get('jwtSecret'),
+				{
+					expiresIn: 360000, //TODO change to 3600 for production
```

Comment:
Do you have an environments file yet? This would be a good addition to that?
Or in the config where the `jwtsecret` is stored?

Reply:
Sounds good, will implement one of these.